CIPs/cip-137.md

```diff
@@ -18,187 +18,227 @@ An API for reading and writing events on Ceramic event streams.
 
 
 ## Motivation
 <!--Motivation is critical for CIPs that want to change the Ceramic protocol. It should clearly explain why the existing protocol specification is inadequate to address the problem that the CIP solves. CIP submissions without sufficient motivation may be rejected outright.-->
-In order to enable databases other than ComposeDB to be built on top of Ceramic, an API to read and write directly on event streams is needed. This also enables developers to get direct access to the event streams underlaying ComposeDB.
+In order to enable services and databases other than ComposeDB to be built on top of Ceramic, an API to read and write directly on event streams is needed. This also enables developers to get direct access to the event streams underlaying ComposeDB.
 
 
 ## Specification
 <!--The technical specification should describe the syntax and semantics of any new feature.-->
 
-This is a work in progress specification that introduces a JSON-RPC interface for interacting directly with event streams on a node.
+This specification introduces a REST API for interacting directly with event streams on a node.
 
 
 
-### `ceramic_subscribe`
+### GET `/ceramic/subscribe/{streamid}`
 
-This method tells the Ceramic node to subscribe to a given interest range of streams. A interest range is defined as `ceramic://*?<sort-key>=<sort-value>`, where *sort-key* describes a field in the header of the InitEvent of a stream, and *sort-value* is the value of this field. For example to subscribe to all MIDs of a particular Model you can use `ceramic://*?model=kjzl6hvfrbw6c82mkud4qs38zl4hd03ifoyg2ksvfjkhuxebfzh3ef89vwvtvrr`.
+This endpoint is called to subscribe to new events in a given interest range. This endpoint uses the node’s local ordering of events within the given range, constructed in the order in which the node received the events. If you want the global ordering of events you can use the *recon* endpoint. 
 
-Once subscribed the Ceramic node will use the Recon protocol to keep in sync with other nodes subscribed to this interest range.
+The range is normally defined as `ceramic://*?<sort-key>=<sort-value>`, where *sort-key* describes a field in the header of the InitEvent of a stream, and *sort-value* is the value of this field. For example to subscribe to all MIDs of a particular Model you can use `ceramic://*?model=kjzl6hvfrbw6c82mkud4qs38zl4hd03ifoyg2ksvfjkhuxebfzh3ef89vwvtvrr`.
 
-Active subscriptions need to be persisted. Or can be stateless somehow? (Nathaniel to share more thoughts here)
+This endpoint will respond as soon as the node knows about new events, given the *eventOffset*, or when *duration* amount of seconds has passed.
```

Comment:
Does this endpoint return a single event or multiple? What determines when we wait for more events? Is that what duration is for? 


For example lets say there is a stream that has events arriving about every 1s. If I call this endpoint with a duration of `5s` when do I get a response? Within 1s when the first event arrives or after 5s when my duration expires? or somewhere in between?  How would I make a call to this endpoint to get events with as little latency as possible? 

There is another challenge I see with using this endpoint in that `eventOffset` is the number of events. If we use number of events as the offset then the server must keep an index to each event so it can start anywhere in the stream and produce the event. In contrast the offset could be a byte offset into `log` file that the server produces. This would mean that when a client asks for new events the server can quickly seek to the correct place where the events are stored. This is how Kafka works as an example  see https://kafka.apache.org/documentation/#impl_reads

Kafka has two design differences that help keep the client simple and the server implementation manageable.

1. Clients asks for events from a byte offset
2. Clients specify a max size in bytes of events they are willing to receive.

The server then responds with as many events as can fit into the max bytes that are immediately available and returns the new offset that the clients can use on their next call. This allows for low latency when events are already present and allows the clients to control how much batching they need. 

Thoughts? Should we use some kind of byte offset instead of an event offset?

Reply:
> If we use a log file with byte offset, we would need to create a new log file for any new subscription (range).

This is the opposite of my intent with using the log file. We do not want the server to have to track any state for clients. Clients should be able to track all of their own state. My thinking is that the server would create a log file per unique sort-value. Then you need an offset per sort-value (i.e. an offset per model).  However that doesn't seem to work as we let ranges span multiple sort-values so you would need multiple offsets. Do we need to reconsider letting clients specify arbitrary ranges? Should we instead force clients to only be able to subscribe to a single sort-value at a time so that the server can manage unique sort-values efficiently? (This is similar to Kafka's partition concept, Kafka has two layers of namespacing, topic name and partition number). If we force a static hierarchy then the server can make tradeoffs accordingly. Need to think on this a bit more...

Reply:
> Should we instead force clients to only be able to subscribe to a single sort-value per subscription so that the server can manage unique sort-values efficiently?

This is not sufficient since we need to be able to subscribe to subset of sort-values if for example a model grows really big. One of the main value props of recon is that nodes can subscribe to arbitrary ranges.

> We do not want the server to have to track any state for clients.

Why is having one log file not considered state for clients? It's not like we would be keeping specific log files for specific clients, but rather for specific ranges.

Reply:
>It's not like we would be keeping specific log files for specific clients, but rather for specific ranges.

There are an enumerable number of possible ranges so this means the server has to keep state based on the parameters of requests from clients. In the happy path there will be lots of reuse of ranges so its not an issue. But it leaves the door open for abuse as all a malicious client has to do is subscribe to lots of different ranges to cause the server to do lots of work. 


I think the core design question I am trying to answer is if we expect server implementations to maintain and index of each event or if they can reduce the overhead of events by storing them without an index.

I am going to write up a more complete discussion around this question.